dali/base/dasess.cpp

```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
         passwordenc.clear();
         sessionToken = 0;
         signature.clear();
+        utcTimeStamp.clear();
+        userTimeStampSignature.clear();
     }
     void serialize(MemoryBuffer &mb)
     {
         mb.append(username).append(passwordenc);
     }
-    void serializeExtra(MemoryBuffer &mb)
+    void serializeSignature(MemoryBuffer &mb)
     {
         if (queryDaliServerVersion().compare("3.14") >= 0)
         {
-            if (signature.isEmpty())
+            if (pDSM == nullptr)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
+
+            //Serialize timestamp and signature of "username;timestamp"
+            //Dali will use this to ensure request came from
+            //valid authenticated user within a reasonable timeframe
+            if (pDSM && pDSM->isDigiSignerConfigured())
             {
-                if (pDSM == nullptr)
-                    pDSM = createDigitalSignatureManagerInstanceFromEnv();
-                if (pDSM && pDSM->isDigiSignerConfigured())
-                    pDSM->digiSign(username, signature);//Set user's digital signature
+                CDateTime now;
+                now.setNow();
+                StringBuffer timeStr;
+                now.getString(timeStr, false);//get UTC timestamp
+                VStringBuffer toSign("%s;%s", username.str(), timeStr.str());
+
+                StringBuffer userTimeStampSignature;
+                pDSM->digiSign(toSign, userTimeStampSignature);//Sign "username;timeStamp"
```

Comment:
queryUserTimeStampSignature() and the 'userTimeStampSignature' member are quite confusing because they depend on the context of the IUserDescriptor.
e.g. if you create a IUserDescriptor locally then call queryUserTimeStampSignature() it is undefined.

These members are just part of the internal serialization protocol that now must vary per request, and don't really belong in IUserDescriptor.

For that reason, I think the date signed signature should serialization should be moved outside of IUserDescriptor and instead be part of the request itself.
That way the IUserDescriptor is constant and clean again and you don't have an undefined member and query method depending on context.

i.e. in the client impl. of getPermissionsLDAP, it can call something like : serializeTimeSignedUserDescriptor(IUserDescriptor *user);

And it can perform the same mangling and digi signing.

Reply:
Agreed, will do

Reply:
@jakesmith Actually now I see that I made the timestamp and user;time signature part of the IUserDescriptor so I would not need to add them as arguments to the getPermissionsLDAPmethod, and the ldapConn->getPermissions() method where the actual authentication takes place.
I will pass them as arguments unless maybe I can add generic key/value pair getter/setter to IUserDescriptor and save them that way???   Tony is wanting me to add the scope as part of the signature (scope;user;timestamp) so this would help
Your opion?

Reply:
It may be convenient to add them the IUserDescriptor implementation(s), but it isn't information that is wanted by users of IUserDescriptor's except by the deserializer and verifier, and the info needs to be different every time (because of date stamp).

So any transient meta info, like date stamp, should be serialized/deserialized when needed and not part of the user descriptor.
If there really need to be extra arguments to getPermissionsLDAP to check LDAP permissions, like date of request, then it is appropriate to add it there.

Reply:
@jakesmith It seems to me that the whole purpose of the IUserDescriptor is to identify and authenticate users.  I can see not serializing/deserializing from the user IUserDescriptor , but it seems that adding generic getProp/setProp is a practical way of adding additional user authentication relics